{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running read alignment with minimap2 on top 20% sequences...\n",
      "Running command: minimap2 -x ava-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf\n",
      "Command 'minimap2 -x ava-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf' took 2.84 seconds to run.\n",
      "Error: [M::mm_idx_gen::0.024*1.20] collected minimizers\n",
      "[M::mm_idx_gen::0.038*1.84] sorted minimizers\n",
      "[M::main::0.038*1.83] loaded/built the index for 1000 target sequence(s)\n",
      "[M::mm_mapopt_update::0.040*1.78] mid_occ = 608\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 1000\n",
      "[M::mm_idx_stat::0.041*1.76] distinct minimizers: 61942 (78.57% are singletons); average occurrences: 3.831; average spacing: 2.966; total length: 703802\n",
      "[M::worker_pipeline::2.794*2.70] mapped 1000 sequences\n",
      "[M::main] Version: 2.22-r1101\n",
      "[M::main] CMD: minimap2 -x ava-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq\n",
      "[M::main] Real time: 2.798 sec; CPU: 7.554 sec; Peak RSS: 0.074 GB\n",
      "\n",
      "Generating consensus sequence with racon on top 20% sequences...\n",
      "Running command: racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta\n",
      "Command 'racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta' took 2.74 seconds to run.\n",
      "Error: [racon::Polisher::initialize] loaded target sequences 0.007375 s\n",
      "[racon::Polisher::initialize] loaded sequences 0.007687 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.591344 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=>                  ] 0.006066 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==>                 ] 0.010291 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===>                ] 0.014380 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====>               ] 0.018647 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=====>              ] 0.022838 s\n",
      "[racon::Polisher::initialize] aligning overlaps [======>             ] 0.027150 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=======>            ] 0.031470 s\n",
      "[racon::Polisher::initialize] aligning overlaps [========>           ] 0.035740 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=========>          ] 0.040056 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==========>         ] 0.044383 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===========>        ] 0.048691 s\n",
      "[racon::Polisher::initialize] aligning overlaps [============>       ] 0.052924 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=============>      ] 0.057114 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==============>     ] 0.061240 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===============>    ] 0.065370 s\n",
      "[racon::Polisher::initialize] aligning overlaps [================>   ] 0.069509 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=================>  ] 0.073435 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==================> ] 0.077361 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===================>] 0.081241 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 0.085870 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.008882 s\n",
      "[racon::Polisher::polish] generating consensus [=>                  ] 1.244877 s\n",
      "[racon::Polisher::polish] generating consensus [==>                 ] 1.796438 s\n",
      "[racon::Polisher::polish] generating consensus [===>                ] 1.923841 s\n",
      "[racon::Polisher::polish] generating consensus [====>               ] 1.938595 s\n",
      "[racon::Polisher::polish] generating consensus [=====>              ] 1.950882 s\n",
      "[racon::Polisher::polish] generating consensus [======>             ] 1.951090 s\n",
      "[racon::Polisher::polish] generating consensus [=======>            ] 1.961098 s\n",
      "[racon::Polisher::polish] generating consensus [========>           ] 1.965399 s\n",
      "[racon::Polisher::polish] generating consensus [=========>          ] 1.965505 s\n",
      "[racon::Polisher::polish] generating consensus [==========>         ] 1.965602 s\n",
      "[racon::Polisher::polish] generating consensus [===========>        ] 1.973723 s\n",
      "[racon::Polisher::polish] generating consensus [============>       ] 1.973975 s\n",
      "[racon::Polisher::polish] generating consensus [=============>      ] 1.978120 s\n",
      "[racon::Polisher::polish] generating consensus [==============>     ] 1.978378 s\n",
      "[racon::Polisher::polish] generating consensus [===============>    ] 1.978670 s\n",
      "[racon::Polisher::polish] generating consensus [================>   ] 1.978860 s\n",
      "[racon::Polisher::polish] generating consensus [=================>  ] 1.981993 s\n",
      "[racon::Polisher::polish] generating consensus [==================> ] 1.986165 s\n",
      "[racon::Polisher::polish] generating consensus [===================>] 1.986288 s\n",
      "[racon::Polisher::polish] generating consensus [====================] 1.990504 s\n",
      "[racon::Polisher::] total = 2.692570 s\n",
      "\n",
      "Multiple sequences found in assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta. Selecting the best alignment...\n",
      "Running read alignment with minimap2 on remaining 80% sequences...\n",
      "Running command: minimap2 -x map-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf\n",
      "Command 'minimap2 -x map-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf' took 0.07 seconds to run.\n",
      "Error: [M::mm_idx_gen::0.001*8.42] collected minimizers\n",
      "[M::mm_idx_gen::0.001*5.40] sorted minimizers\n",
      "[M::main::0.001*5.36] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.001*5.22] mid_occ = 10\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.001*5.09] distinct minimizers: 130 (100.00% are singletons); average occurrences: 1.000; average spacing: 6.031; total length: 784\n",
      "[M::worker_pipeline::0.025*2.25] mapped 4000 sequences\n",
      "[M::main] Version: 2.22-r1101\n",
      "[M::main] CMD: minimap2 -x map-ont assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq\n",
      "[M::main] Real time: 0.026 sec; CPU: 0.057 sec; Peak RSS: 0.006 GB\n",
      "\n",
      "Generating final consensus sequence with racon...\n",
      "Running command: racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_final_consensus.fasta\n",
      "Command 'racon -m 8 -x -6 -g -8 -w 500 assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80.fastq assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_remaining80_reads.paf assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_top20_consensus.fasta > assets/output/rbcL_Qiagen_tomato_5000/rbcL_Qiagen_tomato_5000_final_consensus.fasta' took 9.01 seconds to run.\n",
      "Error: [racon::Polisher::initialize] loaded target sequences 0.000057 s\n",
      "[racon::Polisher::initialize] loaded sequences 0.017786 s\n",
      "[racon::Polisher::initialize] loaded overlaps 0.005886 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=>                  ] 0.025914 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==>                 ] 0.034357 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===>                ] 0.042700 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====>               ] 0.050003 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=====>              ] 0.056697 s\n",
      "[racon::Polisher::initialize] aligning overlaps [======>             ] 0.062778 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=======>            ] 0.068564 s\n",
      "[racon::Polisher::initialize] aligning overlaps [========>           ] 0.073907 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=========>          ] 0.078624 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==========>         ] 0.083080 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===========>        ] 0.087323 s\n",
      "[racon::Polisher::initialize] aligning overlaps [============>       ] 0.091805 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=============>      ] 0.095744 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==============>     ] 0.099726 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===============>    ] 0.103258 s\n",
      "[racon::Polisher::initialize] aligning overlaps [================>   ] 0.106458 s\n",
      "[racon::Polisher::initialize] aligning overlaps [=================>  ] 0.109472 s\n",
      "[racon::Polisher::initialize] aligning overlaps [==================> ] 0.112239 s\n",
      "[racon::Polisher::initialize] aligning overlaps [===================>] 0.115159 s\n",
      "[racon::Polisher::initialize] aligning overlaps [====================] 0.117721 s\n",
      "[racon::Polisher::initialize] transformed data into windows 0.001803 s\n",
      "[racon::Polisher::polish] generated consensus 8.822443 s\n",
      "[racon::Polisher::] total = 8.967894 s\n",
      "\n",
      "Minimap2 alignment took 0.07 seconds.\n",
      "Total Racon iterations took 14.59 seconds.\n",
      "Total time taken for the pipeline: 14.66 seconds.\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import time\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "\n",
    "def compute_alignment_score(seq1, seq2):\n",
    "    alignments = pairwise2.align.globalxx(seq1.seq, seq2.seq, one_alignment_only=True)\n",
    "    if alignments:\n",
    "        alignment = alignments[0]\n",
    "        return alignment.score\n",
    "    else:\n",
    "        return 0  # No alignment found, score is 0\n",
    "\n",
    "def run_command(command):\n",
    "    print(f\"Running command: {command}\")\n",
    "    start_time = time.time()\n",
    "    result = subprocess.run(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)\n",
    "    elapsed_time = time.time() - start_time\n",
    "    print(f\"Command '{command}' took {elapsed_time:.2f} seconds to run.\")\n",
    "    if result.stdout:\n",
    "        print(result.stdout)\n",
    "    if result.stderr:\n",
    "        print(\"Error:\", result.stderr)\n",
    "    return result, elapsed_time\n",
    "\n",
    "def select_best_alignment(sequences, query_sequence):\n",
    "    best_sequence = None\n",
    "    best_score = 0\n",
    "\n",
    "    for sequence in sequences:\n",
    "        alignment_score = compute_alignment_score(sequence, query_sequence)  # Implement this function to compute alignment score\n",
    "        if alignment_score > best_score:\n",
    "            best_score = alignment_score\n",
    "            best_sequence = sequence\n",
    "\n",
    "    return best_sequence\n",
    "\n",
    "def split_fastq(input_path, output_dir, base_name, percentile=20):\n",
    "    # Read sequences and sort by length\n",
    "    sequences = list(SeqIO.parse(input_path, \"fastq\"))\n",
    "    sequences.sort(key=lambda x: len(x), reverse=True)\n",
    "\n",
    "    # Split into top percentile and remaining sequences\n",
    "    split_index = len(sequences) * percentile // 100\n",
    "    top_sequences = sequences[:split_index]\n",
    "    remaining_sequences = sequences[split_index:]\n",
    "\n",
    "    # Write to separate files\n",
    "    top_sequences_path = os.path.join(output_dir, f\"{base_name}_top{percentile}.fastq\")\n",
    "    remaining_sequences_path = os.path.join(output_dir, f\"{base_name}_remaining{100-percentile}.fastq\")\n",
    "    SeqIO.write(top_sequences, top_sequences_path, \"fastq\")\n",
    "    SeqIO.write(remaining_sequences, remaining_sequences_path, \"fastq\")\n",
    "\n",
    "    return top_sequences_path, remaining_sequences_path\n",
    "\n",
    "def main():\n",
    "    input_fastq_filename = \"rbcL_Qiagen_tomato_5000.fastq\"\n",
    "    input_fastq_path = f\"assets/input/{input_fastq_filename}\"\n",
    "    base_name = os.path.splitext(input_fastq_filename)[0]\n",
    "\n",
    "    output_base_dir = \"assets/output\"\n",
    "    output_dir = os.path.join(output_base_dir, base_name)\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    total_time_taken = 0\n",
    "\n",
    "    # Split the file into top 20% and remaining 80%\n",
    "    top_sequences_path, remaining_sequences_path = split_fastq(input_fastq_path, output_dir, base_name, percentile=20)\n",
    "\n",
    "    # Step 1: Align and generate consensus from top 20% sequences\n",
    "    top_paf_path = os.path.join(output_dir, f\"{base_name}_top20_reads.paf\")\n",
    "    top_consensus_path = os.path.join(output_dir, f\"{base_name}_top20_consensus.fasta\")\n",
    "    minimap2_command = f\"minimap2 -x ava-ont {top_sequences_path} {top_sequences_path} > {top_paf_path}\"\n",
    "    print(\"Running read alignment with minimap2 on top 20% sequences...\")\n",
    "    _, minimap2_time = run_command(minimap2_command)\n",
    "    total_time_taken += minimap2_time\n",
    "\n",
    "    racon_command = f\"racon -m 8 -x -6 -g -8 -w 500 {top_sequences_path} {top_paf_path} {top_sequences_path} > {top_consensus_path}\"\n",
    "    print(\"Generating consensus sequence with racon on top 20% sequences...\")\n",
    "    _, racon_time = run_command(racon_command)\n",
    "    total_time_taken += racon_time\n",
    "\n",
    "    # Step 1 bis: If the outputed racon file contains more than one sequence, we take the longest one\n",
    "    consensus_sequences = list(SeqIO.parse(top_consensus_path, \"fasta\"))\n",
    "    if len(consensus_sequences) > 1:\n",
    "        print(f\"Multiple sequences found in {top_consensus_path}. Selecting the best alignment...\")\n",
    "        best_sequence = select_best_alignment(consensus_sequences, consensus_sequences[0])  # Choose the first sequence as the reference\n",
    "        SeqIO.write(best_sequence, top_consensus_path, \"fasta\")\n",
    "\n",
    "    # Step 2: Align the remaining 80% sequences to the top 20% consensus\n",
    "    remaining_paf_path = os.path.join(output_dir, f\"{base_name}_remaining80_reads.paf\")\n",
    "    final_consensus_path = os.path.join(output_dir, f\"{base_name}_final_consensus.fasta\")\n",
    "    minimap2_command = f\"minimap2 -x map-ont {top_consensus_path} {remaining_sequences_path} > {remaining_paf_path}\"\n",
    "    print(\"Running read alignment with minimap2 on remaining 80% sequences...\")\n",
    "    _, minimap2_time = run_command(minimap2_command)\n",
    "    total_time_taken += minimap2_time\n",
    "\n",
    "    # Step 3: Generate the final consensus sequence with racon\n",
    "    racon_command = f\"racon -m 8 -x -6 -g -8 -w 500 {remaining_sequences_path} {remaining_paf_path} {top_consensus_path} > {final_consensus_path}\"\n",
    "    print(\"Generating final consensus sequence with racon...\")\n",
    "    _, racon_time = run_command(racon_command)\n",
    "    total_time_taken += racon_time\n",
    "\n",
    "    # Print out the total time for each step\n",
    "    print(f\"Minimap2 alignment took {minimap2_time:.2f} seconds.\")\n",
    "    print(f\"Total Racon iterations took {total_time_taken - minimap2_time:.2f} seconds.\")\n",
    "\n",
    "    # Print out the total time for the pipeline\n",
    "    print(f\"Total time taken for the pipeline: {total_time_taken:.2f} seconds.\")\n",
    "\n",
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GenoRobotics_bioinfo_consensus",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
